compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -300,60 +301,85 @@ fn negative_impl<'cx, 'tcx>(
     debug!("negative_impl(impl1_def_id={:?}, impl2_def_id={:?})", impl1_def_id, impl2_def_id);
     let tcx = selcx.infcx().tcx;
 
-    // create a parameter environment corresponding to a (placeholder) instantiation of impl1
-    let impl1_env = tcx.param_env(impl1_def_id);
-    let impl1_trait_ref = tcx.impl_trait_ref(impl1_def_id).unwrap();
-
     // Create an infcx, taking the predicates of impl1 as assumptions:
     tcx.infer_ctxt().enter(|infcx| {
-        // Normalize the trait reference. The WF rules ought to ensure
-        // that this always succeeds.
-        let impl1_trait_ref = match traits::fully_normalize(
-            &infcx,
-            FulfillmentContext::new(),
-            ObligationCause::dummy(),
-            impl1_env,
-            impl1_trait_ref,
-        ) {
-            Ok(impl1_trait_ref) => impl1_trait_ref,
-            Err(err) => {
-                bug!("failed to fully normalize {:?}: {:?}", impl1_trait_ref, err);
+        // create a parameter environment corresponding to a (placeholder) instantiation of impl1
+        let impl_env = tcx.param_env(impl1_def_id);
+        let subject1 = match tcx.impl_subject(impl1_def_id) {
+            ImplSubject::Trait(impl1_trait_ref) => {
+                match traits::fully_normalize(
+                    &infcx,
+                    FulfillmentContext::new(),
+                    ObligationCause::dummy(),
+                    impl_env,
+                    impl1_trait_ref,
+                ) {
+                    Ok(impl1_trait_ref) => ImplSubject::Trait(impl1_trait_ref),
+                    Err(err) => {
+                        bug!("failed to fully normalize {:?}: {:?}", impl1_trait_ref, err);
+                    }
+                }
             }
+            subject @ ImplSubject::Inherent(_) => subject,
         };
```

Comment:
Actually I suspect we want this:

```suggestion
        let subject1 = match match traits::fully_normalize(
                &infcx,
                FulfillmentContext::new(),
                ObligationCause::dummy(),
                impl_env,
                tcx.impl_subject(impl1_def_id),
        ) {
                Ok(s) => s,
                Err(err) => bug!("failed to fully normalize {:?}: {:?}", impl1_def_id, err),
        };
```

i.e., it can't hurt to normalize the "inherent" case too, in case there's something like `impl Foo<<vec::IntoIter<u32> as Iterator>::Item>`